src/index/ivf/ivf.cc

```diff
@@ -423,17 +419,36 @@ IvfIndexNode<T>::Search(const DataSet& dataset, const Config& cfg, const BitsetV
                         }
                     }
                 } else if constexpr (std::is_same<T, faiss::IndexIVFFlat>::value) {
-                    auto cur_data = (const float*)data + index * dim;
-                    index_->search_without_codes_thread_safe(1, cur_data, k, distances + offset, ids + offset, nprobe,
-                                                             0, bitset);
+                    auto cur_query = (const float*)data + index * dim;
+                    if (is_cosine) {
+                        auto cur_query_norm = CopyAndNormalizeFloatVec(cur_query, dim);
+                        index_->search_without_codes_thread_safe(1, cur_query_norm.get(), k, distances + offset,
+                                                                 ids + offset, nprobe, 0, bitset);
+                    } else {
+                        index_->search_without_codes_thread_safe(1, cur_query, k, distances + offset, ids + offset,
+                                                                 nprobe, 0, bitset);
+                    }
                 } else if constexpr (std::is_same<T, faiss::IndexScaNN>::value) {
-                    auto cur_data = (const float*)data + index * dim;
+                    auto cur_query = (const float*)data + index * dim;
                     const ScannConfig& scann_cfg = static_cast<const ScannConfig&>(cfg);
-                    index_->search_thread_safe(1, cur_data, k, distances + offset, ids + offset, nprobe,
-                                               scann_cfg.reorder_k.value(), bitset);
+                    if (is_cosine) {
+                        auto cur_query_norm = CopyAndNormalizeFloatVec(cur_query, dim);
+                        index_->search_thread_safe(1, cur_query_norm.get(), k, distances + offset, ids + offset, nprobe,
+                                                   scann_cfg.reorder_k.value(), bitset);
+                    } else {
+                        index_->search_thread_safe(1, cur_query, k, distances + offset, ids + offset, nprobe,
+                                                   scann_cfg.reorder_k.value(), bitset);
+                    }
                 } else {
-                    auto cur_data = (const float*)data + index * dim;
-                    index_->search_thread_safe(1, cur_data, k, distances + offset, ids + offset, nprobe, 0, bitset);
+                    auto cur_query = (const float*)data + index * dim;
+                    if (is_cosine) {
+                        auto cur_query_norm = CopyAndNormalizeFloatVec(cur_query, dim);
```

Comment:
better to rename all this `cur_query_norm` to `cur_normalized_query`, it is confusing to use query_norm which could be considered as a numeric number means norm.